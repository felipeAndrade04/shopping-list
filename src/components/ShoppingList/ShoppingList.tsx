import React from 'react';
import * as S from './ShoppingList.styles';
import { ShoppingListProps } from './ShoppingList.types';
import { ShoppingCard } from '../ShoppingCard';
import { Shopping } from '@app/models';
import { FlatList } from 'react-native';
import { Spacer } from '../Spacer';

export function ShoppingList({ data }: ShoppingListProps) {
  function renderItem({ item }: { item: Shopping }) {
    return <ShoppingCard shopping={item} />;
  }

  return (
    <S.Container>
      <FlatList
        data={data}
        renderItem={renderItem}
        keyExtractor={(item: any) => item.id}
        ItemSeparatorComponent={() => <Spacer dimesion={16} />}
        showsVerticalScrollIndicator={false}
      />
    </S.Container>
  );
}
