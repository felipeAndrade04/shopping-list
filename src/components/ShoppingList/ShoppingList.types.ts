import { Shopping } from '@app/models';

export interface ShoppingListProps {
  data: Shopping[];
}
