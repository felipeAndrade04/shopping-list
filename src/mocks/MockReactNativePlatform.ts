jest.mock('react-native/Libraries/Utilities/Platform', () => {
  // eslint-disable-next-line @typescript-eslint/no-var-requires
  const fs = require('fs');
  const json = JSON.parse(fs.readFileSync('node_modules/react-native/package.json', 'utf8'));
  const platform = jest.requireActual('react-native/Libraries/Utilities/Platform');
  const version = json.version.toString().split('.');
  return {
    ...platform,
    constants: {
      ...platform.constants,
      reactNativeVersion: {
        major: version[0],
        minor: version[1],
        patch: version[2],
      },
    },
  };
});
