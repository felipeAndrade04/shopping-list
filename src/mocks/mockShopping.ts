import { Shopping } from '@app/models';

export const mockedShopping: Shopping = {
  products: [],
  name: 'Lista 8',
  created_at: new Date('01/01/2023'),
  id: '123',
};
