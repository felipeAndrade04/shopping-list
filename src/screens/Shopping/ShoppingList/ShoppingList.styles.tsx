import { Button } from '@app/components';
import styled from 'styled-components/native';

export const Container = styled.View`
  flex: 1;
  justify-content: center;
  background-color: ${({ theme }) => theme.colors.background};
  padding: 24px;
`;

export const ButtonAddProduct = styled(Button)`
  position: absolute;
  bottom: 20px;
  left: 24px;
`;
