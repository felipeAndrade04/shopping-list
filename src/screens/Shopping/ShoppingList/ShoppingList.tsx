import React, { useEffect } from 'react';
import { ShoppingListProps } from './ShoppingList.types';
import * as S from './ShoppingList.styles';
import { Message, ProductCard, Spacer } from '@app/components';
import { useShopping } from '@app/hooks/shopping';
import { FlatList } from 'react-native-gesture-handler';
import { Product } from '@app/models';

export function ShoppingList({ navigation, route }: ShoppingListProps) {
  const { id } = route.params;
  const { getSelectedShopping, updateProducts } = useShopping();
  const selectedShopping = getSelectedShopping(id);

  useEffect(() => {
    navigation.setOptions({ title: selectedShopping?.name ?? '' });
  }, [navigation, selectedShopping]);

  const products = selectedShopping?.products?.sort((a, b) => {
    if (a.name < b.name) {
      return -1;
    }
    if (a.name > b.name) {
      return 1;
    }
    return 0;
  });

  function navigateToAddProduct() {
    navigation.navigate('AddProduct', { id });
  }

  async function onChangeSelectedProduct(selectedProduct: Product, value: boolean) {
    const filteredProducts: Product[] = selectedShopping?.products?.filter(
      (product) => product.id !== selectedProduct.id
    );

    const updatedProducts: Product[] = [
      ...filteredProducts,
      { ...selectedProduct, checked: value },
    ];

    await updateProducts(id, updatedProducts);
  }

  function renderItem({ item }: { item: Product }) {
    return (
      <ProductCard
        key={`${item.id}-${item.checked}-${item.quantity}`}
        isSimpleCard={true}
        product={item}
        isChecked={item.checked}
        changeSelectedProduct={onChangeSelectedProduct}
      />
    );
  }

  return (
    <S.Container>
      {selectedShopping?.products?.length ? (
        <FlatList
          data={products}
          renderItem={renderItem}
          keyExtractor={(item: Product) => item.id}
          ItemSeparatorComponent={() => <Spacer dimesion={12} />}
          showsVerticalScrollIndicator={false}
        />
      ) : (
        <Message
          title="Lista de compras vazia"
          description="Clique no botÃ£o de adicionar para iniciar a criar sua lista"
        />
      )}

      <Spacer dimesion={72} />

      <S.ButtonAddProduct variant="outline" onPress={navigateToAddProduct}>
        + Adicionar produtos
      </S.ButtonAddProduct>
    </S.Container>
  );
}
