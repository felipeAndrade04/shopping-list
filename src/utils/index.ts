export * from './delay';
export * from './date';
