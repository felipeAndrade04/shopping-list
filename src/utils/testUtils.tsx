import React from 'react';
import { authReducer } from '@app/store';
import { ReactNode } from 'react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components/native';
import * as theme from '@app/theme';
import { configureStore } from '@reduxjs/toolkit';

let store = configureStore({
  reducer: {
    auth: authReducer,
  },
});

export const wrapper = ({ children }: { children: ReactNode }) => (
  <ThemeProvider theme={theme}>
    <Provider store={store}>{children}</Provider>
  </ThemeProvider>
);

export function resetStore() {
  store = configureStore({
    reducer: {
      auth: authReducer,
    },
  });
}
