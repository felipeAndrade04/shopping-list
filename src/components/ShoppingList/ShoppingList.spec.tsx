import React from 'react';
import { render } from '@testing-library/react-native';
import { wrapper } from '@app/utils/testUtils';
import { ShoppingList } from './ShoppingList';
import { mockedShopping } from '@app/mocks/mockShopping';

describe('<ShoppingList />', () => {
  it('Should render correctly', () => {
    const tree = render(<ShoppingList data={[mockedShopping]} />, { wrapper }).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
