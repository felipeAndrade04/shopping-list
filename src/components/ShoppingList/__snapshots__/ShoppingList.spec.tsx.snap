// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShoppingList /> Should render correctly 1`] = `
<View
  style={
    [
      {
        "height": "100%",
        "width": "100%",
      },
    ]
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    data={
      [
        {
          "created_at": 2023-01-01T10:00:00.000Z,
          "id": "123",
          "name": "Lista 8",
          "products": [],
          "userId": "123",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 8,
              "opacity": 1,
              "paddingBottom": 12,
              "paddingLeft": 12,
              "paddingRight": 12,
              "paddingTop": 12,
              "shadowColor": "#0d0a0b40",
              "shadowOffset": {
                "height": 4,
                "width": 4,
              },
              "shadowOpacity": 1,
              "shadowRadius": 2,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#0D0A0B",
                  "fontFamily": "Poppins-Bold",
                  "fontSize": 16,
                },
              ]
            }
          >
            Lista 8
          </Text>
          <View
            dimesion={8}
            style={
              [
                {
                  "height": 8,
                  "width": 8,
                },
              ]
            }
          />
          <View
            style={
              [
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                },
              ]
            }
          >
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            >
              <View
                collapsable={false}
                size={25}
                style={
                  {
                    "alignItems": "center",
                    "borderColor": "#ACACAC",
                    "borderRadius": 12.5,
                    "borderWidth": 4.166666666666667,
                    "height": 25,
                    "justifyContent": "center",
                    "transform": [
                      {
                        "rotate": "-45deg",
                      },
                    ],
                    "width": 25,
                  }
                }
              >
                <View
                  collapsable={false}
                  size={25}
                  style={
                    {
                      "borderBottomColor": "transparent",
                      "borderLeftColor": "#0D0A0B",
                      "borderRadius": 12.5,
                      "borderRightColor": "transparent",
                      "borderTopColor": "#0D0A0B",
                      "borderWidth": 4.166666666666667,
                      "height": 25,
                      "position": "absolute",
                      "transform": [
                        {
                          "rotate": "0deg",
                        },
                      ],
                      "width": 25,
                    }
                  }
                />
                <View
                  collapsable={false}
                  size={25}
                  style={
                    {
                      "borderBottomColor": "transparent",
                      "borderLeftColor": "#ACACAC",
                      "borderRadius": 12.5,
                      "borderRightColor": "transparent",
                      "borderTopColor": "#ACACAC",
                      "borderWidth": 4.166666666666667,
                      "height": 25,
                      "position": "absolute",
                      "width": 25,
                    }
                  }
                />
                <View
                  collapsable={false}
                  size={25}
                  style={
                    {
                      "borderBottomColor": "transparent",
                      "borderLeftColor": "#0D0A0B",
                      "borderRadius": 12.5,
                      "borderRightColor": "transparent",
                      "borderTopColor": "#0D0A0B",
                      "borderWidth": 4.166666666666667,
                      "height": 25,
                      "opacity": 0,
                      "position": "absolute",
                      "transform": [
                        {
                          "rotate": "0deg",
                        },
                      ],
                      "width": 25,
                    }
                  }
                />
              </View>
              <View
                dimesion={8}
                style={
                  [
                    {
                      "height": 8,
                      "width": 8,
                    },
                  ]
                }
              />
              <Text
                style={
                  [
                    {
                      "color": "#ACACAC",
                      "fontFamily": "Poppins-Medium",
                      "fontSize": 16,
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "color": "#0D0A0B",
                        "fontFamily": "Poppins-Bold",
                        "fontSize": 16,
                      },
                    ]
                  }
                >
                  0
                </Text>
                /
                0
              </Text>
            </View>
            <Text
              style={
                [
                  {
                    "color": "#ACACAC",
                    "fontFamily": "Poppins-Regular",
                    "fontSize": 14,
                  },
                ]
              }
            >
              0 de jan
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
